### PR TITLE
Optimize Project Instance and expander to reduce memory consumption when building large projects.

### PR DESCRIPTION
**Pull request by pi1024e**
Issue #4199 says a large portion of memory used when creating a snapshot in the Project Instance class is the creation of Sync Objects, so I have optimized this portion to be more efficient.

### CODE REVIEW

**Comment by livarcocc** [comment]
@BenVillalobos for review.

File: src/Build/Instance/ProjectInstance.cs [hunk 1]
```diff
@@ -2669,19 +2660,22 @@ private void CreateEvaluatedIncludeSnapshotIfRequested(bool keepEvaluationCache,
...
                 if (item.DirectMetadata != null)
                 {
-                    directMetadata = new CopyOnWritePropertyDictionary<ProjectMetadataInstance>(item.DirectMetadataCount);
+                    CopyOnWritePropertyDictionary<ProjectMetadataInstance> directMetadata = new CopyOnWritePropertyDictionary<ProjectMetadataInstance>(item.DirectMetadataCount);
                     foreach (ProjectMetadata directMetadatum in item.DirectMetadata)
                     {
                         ProjectMetadataInstance directMetadatumInstance = new ProjectMetadataInstance(directMetadatum);
                         directMetadata.Set(directMetadatumInstance);
                     }
+                    instance = new ProjectItemInstance(this, item.ItemType, ((IItem)item).EvaluatedIncludeEscaped, item.EvaluatedIncludeBeforeWildcardExpansionEscaped, directMetadata, inheritedItemDefinitions, ProjectCollection.Escape(item.Xml.ContainingProject.FullPath));
```
Comment (danmosemsft) [inline src/Build/Instance/ProjectInstance.cs]: Is this change an improvement? I may be missing something but it just increases the size of the code.

**Comment by danmosemsft** [comment]
Thanks for the submission. Can you point to where in this change the allocations are saved? 

Also, do you have data backing up the increase to the string builder size (it may be a good idea, it just needs data, probably the old value was not arbitrarily chosen)

File: src/Shared/ReuseableStringBuilder.cs [hunk 2]
```diff
@@ -254,7 +254,7 @@ private static class ReuseableStringBuilderFactory
             /// because we could otherwise hold a huge builder indefinitely.
             /// This size seems reasonable for MSBuild uses (mostly expression expansion)
             /// </summary>
-            private const int MaxBuilderSize = 1024;
+            private const int MaxBuilderSize = 2048;
```
Comment (BenVillalobos) [inline src/Shared/ReuseableStringBuilder.cs]: Changing MaxBuilderSize is something we're currently experimenting with. Could you provide any data you may have gathered to show how much this reduces memory allocations?

**Comment by AndyGerlicher** [comment]
Team Triage: Thank you for putting the time in to contribute this. Unfortunately as is I don't think we can accept this. It's very difficult to review (many un-related changes) and we're not sure if this actually solves the issue. We would be willing to take another look if you create a new PR and ideally provide some data showing the reduced memory. Thanks again!

**Pull request closed without merging**